File: apps/objc/mousetrap/MousetrapObserverSwarm.m
// Mousetraps application. Copyright (C) 1996 Santa Fe Institute.
// This library is distributed without any warranty; without even the
// implied warranty of merchantability or fitness for a particular purpose.
// See file LICENSE for details and terms of copying.

#import "MousetrapObserverSwarm.h"
#import "MousetrapModelSwarm.h"
#import <collections.h>
#import <swarmobject.h>
#import <analysis.h>

// The MousetrapObserverSwarm is the top-level swarm that
// watches and reports on what's happening in the 
// MousetrapModelSwarm. It is like the lab-bench 
// on which the mousetrap world is located, along with
// the various instruments that we construct to
// monitor that world.

@implementation MousetrapObserverSwarm

// +createBegin: here we set up the default observation parameters.

//    This is a class method (note the "+"), so it adds 
//    functionality to the "factory" method for creating a GUISwarm
//    that is unique to the creation of a MousetrapObserverSwarm.
//    This method is actually invoked in main.m

+createBegin: (id) aZone {
  MousetrapObserverSwarm * obj;
  ProbeMap * probeMap;

  // invoke our superClass createBegin to allocate ourselves.
  // obj is the allocated ObserverSwarm
  
  obj = [super createBegin: aZone];

  // Fill in the relevant parameters (only one, in this case).

  obj->displayFrequency = 1;

  // Also, build a customized probe map. Without a probe map, the default
  // is to show all variables and messages. Here we choose to
  // customize the appearance of the probe, give a nicer interface.

  probeMap = [EmptyProbeMap createBegin: aZone];
  [probeMap setProbedClass: [self class]];
  probeMap = [probeMap createEnd];

  // Add in a bunch of variables, one per simulation parameter
  // and install our custom probeMap into the probeLibrary.

  [probeMap addProbe: [probeLibrary getProbeForVariable: "displayFrequency"
				    inClass: [self class]]];
  [probeLibrary setProbeMap: probeMap For: [self class]];

  return obj;		// We return the newly created ObserverSwarm
}

// createEnd: create objects we know we'll need. In this case, none,
// but you might want to override this.

-createEnd {
  return [super createEnd];
}


// Create the objects used in the display of the model. 
// Here, we create the objects used in the experiment
// Primairly, the Mousetrap model itself, but also the
// various instrumentation that observes the model. 

-buildObjects {
  id modelZone;					// zone for model
  int x, y, size;

  // Let our superClass build any objects it needs to first
  
  [super buildObjects];

  // Then, we create the model that we're actually observing. The
  // model is a subswarm of the observer. We create the model in
  // its own zone, so storage is segregated.
  
  modelZone = [Zone create: [self getZone]];
  mousetrapModelSwarm = [MousetrapModelSwarm create: modelZone];

  // Now create probe objects on the model and ourselves. This gives a
  // simple user interface to let the user change parameters.
  
  [probeDisplayManager createProbeDisplayFor: mousetrapModelSwarm];
  [probeDisplayManager createProbeDisplayFor: self];

  // Instruct the control panel to wait for a button event: we halt here
  // until someone hits a control panel button so the user can get a
  // chance to fill in parameters before the simulation runs
  // If the user hit the quit button here, we quit.

  [controlPanel waitForControlEvent];
  if ([controlPanel getState] == ControlStateQuit)
    return self;

  // OK - the user has specified all the parameters for the simulation.
  // Now we're ready to start.

  // First, let the model swarm build its objects.
  
  [mousetrapModelSwarm buildObjects];

  // create a colormap

  colormap = [XColormap create: globalZone];
  [colormap setColor: 1 ToGrey: 0.3];
  [colormap setColor: 2 ToName: "red"];

  // set up the EZGraph object
  
  triggerGraph = [EZGraph createBegin: [self getZone]];
  [triggerGraph setTitle: "Trigger data vs. time"];
  [triggerGraph setAxisLabelsX: "time" Y: "number triggered"];
  triggerGraph = [triggerGraph createEnd] ;

  // Now, we create two observations to plot on mousetraps with time:
  //    1) The total number of traps triggered so far
  //    2) The current pending trigger events (number of ping-pong
  //           balls still "in the air")

  [triggerGraph createSequence: "Total triggered"
                         withFeedFrom: [mousetrapModelSwarm getStats] 
                          andSelector: M(getNumTriggered)] ;

  [triggerGraph createSequence: "Pending triggers"
                         withFeedFrom: [mousetrapModelSwarm getStats] 
                          andSelector: M(getNumBalls)] ;

  // Next, create a 2d window to display the mousetrap world
  // and  set its size, zoom factor, title.

  displayWindow = [ZoomRaster create: globalZone];
  [displayWindow setColormap: colormap];
  [displayWindow setZoomFactor: 6];
  [displayWindow setWidth: [mousetrapModelSwarm getGridSize]
		 Height: [mousetrapModelSwarm getGridSize]];
  [displayWindow setWindowTitle: "Mousetrap World"];

  // draw all the mousetraps 

  size = [mousetrapModelSwarm getGridSize];
  for (x = 0; x < size; x++)
    for (y = 0; y < size; y++) {
      Mousetrap * trap;
      trap = [mousetrapModelSwarm getMousetrapAtX: x Y: y];
      if (trap) {
	[displayWindow drawPointX: x Y: y Color: 1];
	[trap setDisplayWidget: displayWindow];
      }
    }
	
  [displayWindow drawSelf];
  [displayWindow pack];

  // Also create an Object2dDisplay: this object is just used
  // to receive and handle probes on mousetraps.

  // Note that we currently have to set a DisplayMessage, but
  // we don't use it - this will get fixed, as Object2dDisplays 
  // have other uses besides "displaying".
  // Compare with Heatbugs, in which we have a collection of heatbugs
  // to display all at once. In mousetraps, traps display themselves
  // when they are triggered (as well as when they are initialized, above).
  // This is a little unclean as is, but the main point of this
  // app is to illustrate dynamic scheduling, so we'll let it slide
  // for now...

  mousetrapDisplay = [Object2dDisplay createBegin: [self getZone]];
  [mousetrapDisplay setDisplayWidget: displayWindow];
  [mousetrapDisplay setDiscrete2dToDisplay: [mousetrapModelSwarm getWorld]];
  [mousetrapDisplay setDisplayMessage: M(drawSelfOn:)];   // draw method
  mousetrapDisplay = [mousetrapDisplay createEnd];

  // And tell the displayWindow to send mouse clicks to the mousetrapDisplay
  // this allows the user to right-click on the display to probe the traps.

  [displayWindow setButton: ButtonRight 
		 Client:    mousetrapDisplay
                 Message:   M(makeProbeAtX:Y:)];

  // All done - we're ready to build a schedule and go
  
  return self;
}  

// Create the actions necessary for the simulation. This is where
// the schedule is built (but not run!)

// Here we create a display schedule - this is used to display the
// state of the world and check for user input. This schedule should
// be thought of as independent from the model - in particular, you
// will also want to run the model without any display.


-buildActions {

  // First, let our superclass build any actions

  [super buildActions];

  // Then, let our model swarm build its own schedule
  
  [mousetrapModelSwarm buildActions];

  // Create an ActionGroup for display: a bunch of things that occur in
  // a specific order, but in the same step of simulation time. Some of 
  // these actions could be executed in parallel, but we don't explicitly
  // notate that here.

  displayActions = [ActionGroup create: [self getZone]];

  // Schedule up the methods to draw the display of the world
  //  and to update the graph of the statistics we're keeping

  [displayActions createActionTo: displayWindow       message: M(drawSelf)];
  [displayActions createActionTo: triggerGraph	      message: M(step)];

  // Schedule the update of the probe display

  [displayActions createActionTo: probeDisplayManager message: M(update)];

  // Check to see if the simulation has ended (all the balls have landed)

  [displayActions createActionTo: self                message: M(checkToStop)];

  // Finally, schedule an update for the whole user interface code.
  // This is crucial: without this, no graphics update and the control
  // panel will be dead. It's best to put it at the end of the display schedule

  [displayActions createActionTo: controlPanel        message: M(doTkEvents)];

  // And the display schedule. Note the repeat interval is set from our
  // own Swarm data structure. Display is frequently the slowest part of a
  // simulation, so redrawing less frequently can be a help.
  
  displaySchedule = [Schedule createBegin: [self getZone]];
  [displaySchedule setRepeatInterval: displayFrequency];
  displaySchedule = [displaySchedule createEnd];
  [displaySchedule at: 0 createAction: displayActions];

  return self;
}  

// activateIn: - activate the schedules so they're ready to run.
// The swarmContext argument has to do with what we were activated *in*.
// Typically the ObserverSwarm is the top-level Swarm, so it's activated
// in "nil". But other Swarms and Schedules and such will be activated
// inside of us.

-activateIn: (id) swarmContext {

  // First, activate ourselves (just pass along the context).

  [super activateIn: swarmContext];

  // Activate the model swarm in ourselves. The model swarm is a
  // subswarm of the observer swarm.

  [mousetrapModelSwarm activateIn: self];

  // Now activate our schedule in ourselves. This arranges for the
  // execution of the schedule we built.

  [displaySchedule activateIn: self];

  // Activate returns the swarm activity - the thing that's ready to run,
  // which is the activity tree rooted in "nil" that we've just added
  // ourselves onto as new leaves
  
  return [self getSwarmActivity];
}

// monitor method - if all the balls have landed, time to quit!

-checkToStop {
  if ([[mousetrapModelSwarm getStats] getNumBalls] == 0) {
    printf("All the balls have landed!\n");

    [controlPanel setStateStopped];
  }

  return self;
}

@end

File: apps/java/jmousetrap/MousetrapObserverSwarm.java
import swarm.*;
import swarm.activity.*;
import swarm.objectbase.*;
import swarm.simtoolsgui.*;
import swarm.random.*;
import swarm.defobj.*;
import swarm.gui.*;
import swarm.analysis.*;
import swarm.space.*;
import swarm.random.*;
public class MousetrapObserverSwarmImpl extends GUISwarmImpl
{
    public  SwarmEnvironment se;
    public int displayFrequency;
    public ScheduleImpl displaySchedule;
    public MousetrapModelSwarmImpl mousetrapModelSwarm;
  
    public ColormapImpl colormap;
    public  ZoomRasterImpl displayWindow;
    public EZGraphImpl triggerGraph;

    public Object2dDisplayImpl mousetrapDisplay;

    public ActivityControlImpl observerActCont;
    public  ActionGroupImpl displayActions;
    public void nag (String s)
    {
        System.out.println (this.getClass().getName() + ":" + s);
        System.out.flush ();
    }
    public void noMethod (Object a)
    {
    }
    public Object _setupMousetraps_ ()
    {
        int x, y, size;
    
        size = mousetrapModelSwarm.getGridSize();
        for (x = 0; x < size; x++)
            for (y = 0; y < size; y++) {
                Mousetrap trap = mousetrapModelSwarm.getMousetrapAtX$Y (x, y);
                if (trap != null)
                    {
                        if (displayWindow != null)
                            displayWindow.drawPointX$Y$Color (x, y, (byte) 1);
                        trap.setDisplayWidget (displayWindow);
                    }
            }
        return this;
    }
    
    public Object _displayWindowDeath_ (Object caller)
    {
        displayWindow.drop();
        displayWindow = null;
        this._setupMousetraps_();
        return this;
    }

    public Object _scheduleItemCanvasDeath_ (Object caller)
    {

        this._setupMousetraps_();
        return this;
    }



    public Object buildObjects ()
    {
        ZoneImpl modelZone = new ZoneImpl ();
        ZoneCImpl imodelZone = new ZoneCImpl (modelZone);
        ZoomRasterCImpl izr;
        Object2dDisplayCImpl io2d;
        MousetrapModelSwarmCImpl immswarm;
        ColormapCImpl icmp;
        EZGraphCImpl iezg;
        super.buildObjects();
        //    nag ("zone create");
        modelZone = (ZoneImpl) imodelZone.create(this.getZone());
        //    nag ("mousetrapmodelswarm");
        mousetrapModelSwarm = new MousetrapModelSwarmImpl ();
        immswarm = new MousetrapModelSwarmCImpl(mousetrapModelSwarm);
        immswarm.se = se;
        //    nag ("create");
        mousetrapModelSwarm = 
            (MousetrapModelSwarmImpl) immswarm.create (this.getZone());
        mousetrapModelSwarm.se = se;
        //    nag ("createArchivedProbeDisplay (mousetrapModelSwarm)");
        se.createArchivedProbeDisplay (mousetrapModelSwarm);
        //    nag ("createArchivedProbeDisplay (this)");
        se.createArchivedProbeDisplay (this);
        //    nag ("Action cache");
        ((ActionCacheImpl)getActionCache()).waitForControlEvent();
        //    nag ("control panel");
        if (((ControlPanelImpl)this.getControlPanel()).getState() == se.ControlStateQuit)
            return this;
        //    nag ("mousetrapModel swarm buildObjects");
        mousetrapModelSwarm.buildObjects ();
        //    nag ("colormpa");
        colormap = new ColormapImpl ();
        icmp = new ColormapCImpl (colormap);
    
        colormap = (ColormapImpl) icmp.create (se.globalZone);
    
        colormap.setColor$ToGrey ((byte) 1, 0.3);
        colormap.setColor$ToName ((byte) 2, "red");
    
        triggerGraph = new EZGraphImpl ();
        iezg = new EZGraphCImpl (triggerGraph);

        iezg.createBegin(this.getZone());
        se.setWindowGeometryRecordName (iezg);
        iezg.setTitle ("Tirgger data vs. time");
        iezg.setAxisLabelsX$Y ("time", "number triggered");
        triggerGraph = (EZGraphImpl) iezg.createEnd();

        try {
      
            Selector slct1, slct2;
            slct1 = new Selector (mousetrapModelSwarm.getStats().getClass(),
                                  "getNumTriggered", false);
            triggerGraph.createSequence$withFeedFrom$andSelector ("Total triggered",
                                                                  mousetrapModelSwarm.getStats(), slct1);
            //      nag ("first part");
            slct2 = new Selector (mousetrapModelSwarm.getStats().getClass(),
                                  "getNumBalls", false);
            //      nag ("done selector");
            triggerGraph.createSequence$withFeedFrom$andSelector ("Pending triggers",
                                                                  mousetrapModelSwarm.getStats(),
                                                                  slct2);
            //      nag ("done sequence");
				   
        } 
        catch (Exception e) 
            { 
                System.out.println ("Exception trigger : " + e.getMessage());
      
            }
        displayWindow = new ZoomRasterImpl ();
        //    nag ("display window");
        izr = new ZoomRasterCImpl (displayWindow);
        izr.createBegin (se.globalZone);
        //    nag (" izr");
        se.setWindowGeometryRecordName (izr);
        //    nag (" se.setWindowGeometryRecordName (izr);");
        displayWindow = (ZoomRasterImpl) izr.createEnd ();
        //    nag ("done display Window createEnd");
        try 
            {
                Selector slct;
                slct = new Selector (this.getClass(), "_displayWindowDeath_", false);
	
                displayWindow.enableDestroyNotification$notificationMethod (this, 
                                                                            slct);
            } catch (Exception e)
                {
                    System.out.println ("Exception display window: " + e.getMessage());
                }
        displayWindow.setColormap (colormap);
        displayWindow.setZoomFactor (6);
        displayWindow.setWidth$Height (mousetrapModelSwarm.getGridSize (),
                                       mousetrapModelSwarm.getGridSize ());
        displayWindow.setWindowTitle ("Mousetrap World");
        this._setupMousetraps_();
        displayWindow.pack();

        mousetrapDisplay = new Object2dDisplayImpl ();
        io2d = new Object2dDisplayCImpl (mousetrapDisplay);
    
        io2d.createBegin (this.getZone());
        io2d.setDisplayWidget (displayWindow);
        io2d.setDiscrete2dToDisplay (mousetrapModelSwarm.getWorld());
        try {
            /// strange noMehtod:
            //      nag ("strange method");
            Selector slct = new Selector (Class.forName ("Mousetrap"), "noMethod", false);
            io2d.setDisplayMessage (slct);
    
        }
        catch (Exception e) { 
            System.out.println ("Exception no mehtod:" + e.getMessage());
        }
        //    nag ("$$");
        mousetrapDisplay = (Object2dDisplayImpl) io2d.createEnd();
        /*
          try
          {
          Selector slct = new Selector (mousetrapDisplay.getClass(),
          "makeProbeAtX$Y$", true);
          displayWindow.setButton$Client$Message (3, mousetrapDisplay,
          slct);
          } catch (Exception e)
          {
	  System.out.println ("Exception make probe: " + e.getMessage());
	  }*/
        return this;
    }
  
    public Object _update_ ()
    {
        if (displayWindow != null)
            displayWindow.drawSelf ();
        return this;
    }
   
    public Object buildActions ()
    {
        ActionGroupCImpl iag;
        ScheduleCImpl isch;
        Selector slct;
        super.buildActions();
        //    nag ("mousetrap model actions");
        mousetrapModelSwarm.buildActions();
    
        displayActions = new ActionGroupImpl();
    
        iag = new ActionGroupCImpl (displayActions);
        displayActions = (ActionGroupImpl) iag.create (this.getZone());
        //    nag ("display actions done");
        displaySchedule = new ScheduleImpl();
        isch = new ScheduleCImpl (displaySchedule);
    
        isch.createBegin (this.getZone());
        isch.setRepeatInterval (displayFrequency);
        displaySchedule = (ScheduleImpl) isch.createEnd();
        //    nag ("display schedule");
        try
            {
	
                slct = new Selector (this.getClass(), "_update_", false);
                displayActions.createActionTo$message (this, slct);
	
                slct = new Selector (triggerGraph.getClass(), "step", true);
                displayActions.createActionTo$message (triggerGraph, slct);
                slct = new Selector (se.probeDisplayManager.getClass(), "update", 
                                     true);
                displayActions.createActionTo$message (se.probeDisplayManager, slct);
	
                slct = new Selector (this.getClass (), "checkToStop", true);
                displayActions.createActionTo$message (this, slct);

                slct = new Selector (this.getActionCache().getClass (), 
                                     "doTkEvents", true);
                displayActions.createActionTo$message (this.getActionCache(), 
                                                       slct);
	
                displaySchedule.at$createAction (0, displayActions);
                //	nag ("display schedule done");
            } catch (Exception e) 
                {
                    System.out.println ("Exception doTkE: " + e.getMessage());
                }
        return this;
    }

    public Object activateIn (Object swarmContext)
    {
        ActivityControlCImpl iac;
        super.activateIn (swarmContext);
        //    nag ("super");
        mousetrapModelSwarm.activateIn (this);
        //    nag ("mousetrapmodel");
        displaySchedule.activateIn (this);
        //    nag ("display scheudle");
        observerActCont = new ActivityControlImpl ();
        iac = new ActivityControlCImpl (observerActCont);
        iac.createBegin (this.getZone());
        nag ("control create begin");
        iac.attachToActivity (this.getActivity());
        nag ("attach to activity"); 
    
        observerActCont = (ActivityControlImpl) iac.createEnd ();
        nag ("control create end");
    
        observerActCont.setDisplayName ("Observer Swarm Controller");

        nag ("archived probe display");
        se.createArchivedProbeDisplay (observerActCont);
        nag ("probe display");
        return this.getActivity();
    
    }

    public Object checkToStop ()
    {
        if (((MousetrapStatistics)mousetrapModelSwarm.getStats()).getNumBalls() == 0)
            {
                System.out.println ("All balls have landed!\n");
                ((ControlPanelImpl)this.getControlPanel()).setStateStopped();
            }
        return this;
    }
			
}
